src/cmd/launch.rs

```diff
@@ -22,3 +126,228 @@ impl LaunchStruct {
         LaunchStruct
     }
 }
+
+/// デバックコンテナを起動するために必要なディレクトリ群を初期化
+///
+/// - `~/.injesh/containers/<CONTAINER_NAME>/upper`
+///
+///     デバックコンテナ起動前の`/var/lib/docker/overlay2/<HASH_ID>/upper`を保存しておくためのディレクトリ
+///
+/// - `~/.injesh/containers/<CONTAINER_NAME>/setting.yaml`
+///
+///     デバックコンテナの`rootfs`やデバックコンテナ作成後に実行するコマンドなどを保存する設定ファイル
+fn initialize_setting<DO: Downloader>(
+    launch: &command::Launch<DO>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let user = user::User::new()?;
+    let dcontainer_base = format!("{}/{}", user.containers(), launch.name());
+
+    // rootfsの種類などが記載された設定ファイルsetting.yamlを~/.injesh/containers/に作成する
+    match Path::new(&dcontainer_base).exists() {
+        true => return Err(Error::AlreadyExists)?,
+        false => {
+            fs::create_dir_all(format!("{}/upper", &dcontainer_base))?;
+            let mut setting_file = fs::File::create(format!("{}/setting.yaml", &dcontainer_base))?;
+            // TODO: 設定ファイルの形式を決めてない
+            setting_file.write_all(b"content of setting.yaml")?;
+        }
+    }
+
+    // /var/lib/docker/overlay2/<HASH_ID>/upperを~/.injesh/containers/<hoge>/upperに対してコピーする
+    copy_dir_recursively(
+        launch.target_container().upperdir(),
+        format!("{}/upper", &dcontainer_base),
+    )?;
+
+    Ok(())
+}
+
+/// ディレクトリを再起的にコピーする関数
+/// ```ignore
+/// copy_dir_recursively("/path/to/from", "/path/to/to")
+/// ```
+fn copy_dir_recursively<P, Q>(from: P, to: Q) -> Result<u64, Box<dyn std::error::Error>>
+where
+    P: AsRef<OsStr> + AsRef<Path>,
+    Q: AsRef<OsStr> + AsRef<Path>,
+{
+    if !Path::new(&from).is_dir() {
+        Err(Error::InputValue)?
+    }
+    if !Path::new(&to).is_dir() {
+        Err(Error::InputValue)?
+    }
+
+    let mut copy_num = 0;
+    for entry_result in fs::read_dir(from)? {
+        let entry = entry_result?;
+        let to_name = Path::new(&to).to_str().ok_or(Error::InputValue)?;
+        let to_path_string = format!(
+            "{}/{}",
+            to_name,
+            entry
+                .file_name()
+                .as_os_str()
+                .to_str()
+                .ok_or(Error::InputValue)?
+        );
+        let to_path = Path::new(&to_path_string);
+
+        if entry.file_type()?.is_dir() {
+            fs::create_dir(&to_path)?;
+            copy_dir_recursively(entry.path(), &to_path)?;
+        } else {
+            fs::copy(entry.path(), &to_path)?;
+        }
+    }
+    Ok(1)
```

Comment:
- 繰り返し内で絶対に変わらない値を初期化するのを解消するためforの外に出した
- 一旦 &str にしてからPathに変換しているようだが、[DirEntryはPathBufに直接変更できる](https://doc.rust-lang.org/std/fs/struct.DirEntry.html#method.path)ので、Pathにそろえてから`join`でつなげてやって簡潔にした
- u64が返っても仕事しないので、返り値の型を`Ok(())`に変更した

```suggestion
fn copy_dir_recursively<P, Q>(from: P, to: Q) -> Result<(), Box<dyn std::error::Error>>
where
    P: AsRef<OsStr> + AsRef<Path>,
    Q: AsRef<OsStr> + AsRef<Path>,
{
    if !Path::new(&from).is_dir() {
        Err(Error::InputValue)?
    }
    if !Path::new(&to).is_dir() {
        Err(Error::InputValue)?
    }

    let to_name = Path::new(&to).to_str().ok_or(Error::InputValue)?;
    for entry_result in fs::read_dir(from)? {
        let entry = entry_result?;

        let to_path = Path::new(to_name).join(entry.path().file_name().ok_or(Error::InputValue)?);
        
        if entry.file_type()?.is_dir() {
            fs::create_dir(&to_path)?;
            copy_dir_recursively(entry.path(), &to_path)?;
        } else {
            fs::copy(entry.path(), &to_path)?;
        }
    }

    Ok(())
```